src/cargo/ops/cargo_add/mod.rs

```diff
@@ -124,11 +126,71 @@ pub fn add(workspace: &Workspace<'_>, options: &AddOptions<'_>) -> CargoResult<(
                 inherited_features.iter().map(|s| s.as_str()).collect();
             unknown_features.extend(inherited_features.difference(&available_features).copied());
         }
+
         unknown_features.sort();
+
+        let mut activated: IndexSet<_> =
+            dep.features.iter().flatten().map(|s| s.as_str()).collect();
+        if dep.default_features().unwrap_or(true) {
+            activated.insert("default");
+        }
         if !unknown_features.is_empty() {
-            anyhow::bail!("unrecognized features: {unknown_features:?}");
+            let (activated, deactivated) = dep.features();
+            let deactivated = deactivated
+                .iter()
+                .filter(|f| !unknown_features.contains(f))
+                .collect::<Vec<_>>();
+            let activated = activated
+                .iter()
+                .filter(|f| !unknown_features.contains(f))
+                .collect::<Vec<_>>();
```

Comment:
Why is this needed?

Reply:
if this is not included, the unknown features are included in the output.

```
$ cargo add tokio --features sync,net,rt,rr,dd
[...snip]
error: unrecognized features for crate tokio: dd, rr
disabled features:
    bytes, fs, full, io-std, io-util, macros, memchr, mio, num_cpus, parking_lot
    process, rt-multi-thread, signal, signal-hook-registry, stats, test-util, time
    tokio-macros, tracing, winapi
enabled features:
    dd, libc, net, once_cell, rr, rt, socket2, sync
```

I guess the filter on deactivated is not needed, is it possible to remove features with cargo add or will that come with the eventual `cargo remove/rm` merge?

Reply:
I switched this to use `retain` instead

Reply:
Oh because we've already merged them into the `DependencyUI`.  Probably deserves a comment.